### PR TITLE
fix(bazel): Fix appending bazel-out to gitignore

### PR DESCRIPTION
**Pull request by kyliau**
The current implementation appends '/bazel-out' to the wrong location
in the .gitignore file. This PR fixes that and adds a stricter test.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/bazel/src/schematics/ng-new/index.ts [hunk 1]
```diff
@@ -85,26 +85,25 @@ function addDevAndProdMainForAot(options: Schema) {
...
       return host;
     }
-    const gitIgnoreContent = host.read(gitignore);
-    if (!gitIgnoreContent) {
-      throw new Error('Failed to read .gitignore content');
-    }
-
-    if (gitIgnoreContent.includes('/bazel-out\n')) {
+    const gitIgnoreContent = host.read(gitignore).toString();
+    if (gitIgnoreContent.includes('/bazel-out')) {
```
Comment (mgechev) [inline packages/bazel/src/schematics/ng-new/index.ts]: If we drop the `\n` suffix `.includes('/bazel-out-foo')` will also return true; why do we have to remove it?

**Review by IgorMinar — CHANGES_REQUESTED** [review]

File: packages/bazel/src/schematics/ng-new/index.ts (hunk 1 above)
Comment (kyliau) [reply packages/bazel/src/schematics/ng-new/index.ts]: Modified the code to check for start and end lines. The gitignore file is generated by the `ng new` template, and in schematics all templates have '\r' endings only.

File: packages/bazel/src/schematics/ng-new/index_spec.ts [hunk 2]
```diff
@@ -94,7 +94,7 @@ describe('Ng-new Schematic', () => {
     const {files} = host;
     expect(files).toContain('/demo/.gitignore');
     const content = host.readContent('/demo/.gitignore');
-    expect(content).toMatch('/bazel-out');
+    expect(content).toMatch('# compiled output\n/bazel-out');
```
[the remark this answers is not in the document]
Comment (kyliau) [reply packages/bazel/src/schematics/ng-new/index_spec.ts]: done

File: packages/bazel/src/schematics/ng-new/index.ts [hunk 3]
```diff
@@ -85,26 +85,25 @@ function addDevAndProdMainForAot(options: Schema) {
   };
 }
 
-function overwriteGitignore(options: Schema) {
+/**
+ * Apend '/bazel-out' to the gitignore file.
```
[the remark this answers is not in the document]
Comment (kyliau) [reply packages/bazel/src/schematics/ng-new/index.ts]: thanks for the correction.

File: packages/bazel/src/schematics/ng-new/index.ts [hunk 4]
```diff
@@ -85,26 +85,25 @@ function addDevAndProdMainForAot(options: Schema) {
...
       return host;
     }
-    const lines = gitIgnoreContent.toString().split(/\n/g);
+    const compiledOutput = '# compiled output\n';
+    const index = gitIgnoreContent.indexOf(compiledOutput);
+    const insertionIndex = index >= 0 ? index + compiledOutput.length : gitIgnoreContent.length;
     const recorder = host.beginUpdate(gitignore);
-    const compileOutput = lines.findIndex((line: string) => line === '# compiled output');
-    recorder.insertRight(compileOutput, '\n/bazel-out');
+    recorder.insertRight(insertionIndex, '/bazel-out\n');
```
Comment (gkalpak) [inline packages/bazel/src/schematics/ng-new/index.ts]: This assumes that the file ends with a newline. Is this guaranteed to be the case?

Comment (kyliau) [reply packages/bazel/src/schematics/ng-new/index.ts]: Yes, please see the comment in https://github.com/angular/angular/pull/28385#discussion_r251542430

tldr:  The gitignore file is generated by the ng new template, and in schematics all templates have '\r' endings only.

Comment (kyliau) [reply packages/bazel/src/schematics/ng-new/index.ts]: ah, yes, I meant `\n` line endings.

1. The templates end each line with a line ending
Yes, because this code is run immediately after the original `ng new` is run, it's effectively modifying this template [file](https://github.com/angular/angular-cli/blob/master/packages/schematics/angular/workspace/files/__dot__gitignore.template).

2. There is no way the file could have been updated since its creation
Same argument as above

i.e. This is an extension to the original `ng new` and it's modifying the template output of the schematic.

Comment (gkalpak) [reply packages/bazel/src/schematics/ng-new/index.ts]: Hm...In that case, why not assume that `# compiled output` will always exist?

Comment (kyliau) [reply packages/bazel/src/schematics/ng-new/index.ts]: that is party to make the implementation more robust. In case upstream `ng new` changes the template, this implementation would just append the line to the end of the file.

**Comment by kyliau** [comment]
Changes have been incorporated into https://github.com/angular/angular/pull/28436 as part of a major refactoring of `@angular/bazel`.
